File: pbark/pbark/parser/collection/smart_ordinal_parser.py
from __future__ import annotations

_VOCAB: dict[str, int] = {}

_ONES = [
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen",
    "seventeen", "eighteen", "nineteen",
]
_ORDINALS = [
    "zeroth", "first", "second", "third", "fourth", "fifth", "sixth", "seventh",
    "eighth", "ninth", "tenth", "eleventh", "twelfth", "thirteenth", "fourteenth",
    "fifteenth", "sixteen-th", "seventeenth", "eighteenth", "nineteenth",
]
for i, (one, ord_) in enumerate(zip(_ONES, _ORDINALS)):
    _VOCAB[one] = i
    _VOCAB[ord_] = i

for word, val in [
    ("twenty", 20), ("twentieth", 20), ("thirty", 30), ("thirtieth", 30),
    ("forty", 40), ("fortieth", 40), ("fifty", 50), ("fiftieth", 50),
    ("sixty", 60), ("sixtieth", 60), ("seventy", 70), ("seventieth", 70),
    ("eighty", 80), ("eightieth", 80), ("ninety", 90), ("ninetieth", 90),
    ("hundred", 100), ("hundredth", 100), ("thousand", 1000), ("thousandth", 1000),
    ("million", 1_000_000), ("millionth", 1_000_000),
]:
    _VOCAB[word] = val


def parse_word_to_number(input_word: str) -> int:
    if input_word is None or not input_word.strip():
        return 0
    clean = input_word.strip().lower()
    if clean == "last":
        return -1
    tokens = clean.replace("-", " ").split()
    total_sum = 0
    current_section = 0
    for token in tokens:
        if token == "and":
            continue
        if token not in _VOCAB:
            continue
        value = _VOCAB[token]
        if value == 100:
            current_section *= value
        elif value >= 1000:
            current_section *= value
            total_sum += current_section
            current_section = 0
        else:
            current_section += value
    return total_sum + current_section

File: src/main/java/dev/klomptech/jbark/parser/collection/SmartOrdinalParser.java
package dev.klomptech.jbark.parser.collection;

import java.util.HashMap;
import java.util.Map;

// Remind me again why I had to overdo this and not just limit to zero to ten?
public final class SmartOrdinalParser {

  private SmartOrdinalParser() {}

  private static final Map<String, Integer> VOCAB = new HashMap<>();

  static {
    // Base numbers and their standard/ordinal variations mapped to single primitives
    String[] ones = {
      "zero",
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen"
    };
    String[] ordinals = {
      "zeroth",
      "first",
      "second",
      "third",
      "fourth",
      "fifth",
      "sixth",
      "seventh",
      "eighth",
      "ninth",
      "tenth",
      "eleventh",
      "twelfth",
      "thirteenth",
      "fourteenth",
      "fifteenth",
      "sixteen-th",
      "seventeenth",
      "eighteenth",
      "nineteenth"
    };
    for (int i = 0; i < ones.length; i++) {
      VOCAB.put(ones[i], i);
      VOCAB.put(ordinals[i], i);
    }
    // Tens rules
    VOCAB.put("twenty", 20);
    VOCAB.put("twentieth", 20);
    VOCAB.put("thirty", 30);
    VOCAB.put("thirtieth", 30);
    VOCAB.put("forty", 40);
    VOCAB.put("fortieth", 40);
    VOCAB.put("fifty", 50);
    VOCAB.put("fiftieth", 50);
    VOCAB.put("sixty", 60);
    VOCAB.put("sixtieth", 60);
    VOCAB.put("seventy", 70);
    VOCAB.put("seventieth", 70);
    VOCAB.put("eighty", 80);
    VOCAB.put("eightieth", 80);
    VOCAB.put("ninety", 90);
    VOCAB.put("ninetieth", 90);

    // Scales
    VOCAB.put("hundred", 100);
    VOCAB.put("hundredth", 100);
    VOCAB.put("thousand", 1000);
    VOCAB.put("thousandth", 1000);
    VOCAB.put("million", 1000000);
    VOCAB.put("millionth", 1000000);
  }

  public static int parseWordToNumber(final String input) {
    if (input == null || input.isBlank()) return 0;
    // Handle custom keywords like last
    String clean = input.trim().toLowerCase();
    if (clean.equals("last")) return -1;
    // Split text by spaces, hyphens, and remove "and" bridges
    String[] tokens = clean.split("[\\s-]+");
    int totalSum = 0;
    int currentSection = 0;
    for (String token : tokens) {
      if (token.equals("and")) continue;
      if (!VOCAB.containsKey(token)) continue; // Ignore unrecognized tokens
      int value = VOCAB.get(token);
      if (value == 100) {
        // Hundreds scale applies to the immediate group
        currentSection *= value;
      } else if (value >= 1000) {
        // Major scales lock in the current segment and multiply it
        currentSection *= value;
        totalSum += currentSection;
        currentSection = 0;
      } else {
        // Add units and tens directly
        currentSection += value;
      }
    }
    return totalSum + currentSection;
  }
}
